Source/WebCore/rendering/RenderBlockFlow.cpp

```diff
@@ -728,21 +729,20 @@ void RenderBlockFlow::layoutBlockChildren(bool relayoutChildren, LayoutUnit& max
     // The margin struct caches all our current margin collapsing state.
     MarginInfo marginInfo(*this, beforeEdge, afterEdge);
 
-    auto hasMarginTrimState = false;
-    auto updateMarginTrimStateIfNeeded = [&] {
-        auto containingBlockTrimmingState = layoutState->blockStartTrimming();
+    bool blockStartTrimmingFromContainingBlock = layoutState->blockStartTrimming();
+    bool newBlockStartTrimmingForSubtree = [&] {
         if (style().marginTrim().contains(MarginTrimType::BlockStart))
-            layoutState->pushBlockStartTrimming(true);
-        else if (!marginInfo.canCollapseMarginBeforeWithChildren() && containingBlockTrimmingState)
-            layoutState->pushBlockStartTrimming(false);
-        else if (marginInfo.canCollapseMarginBeforeWithChildren() && containingBlockTrimmingState)
-            layoutState->pushBlockStartTrimming(containingBlockTrimmingState.value());
-        else
-            return;
-        hasMarginTrimState = true;
-    };
+            return true;
+        if (!marginInfo.canCollapseMarginBeforeWithChildren() && blockStartTrimmingFromContainingBlock)
+            return false;
+        return blockStartTrimmingFromContainingBlock;
+    }();
+
+    layoutState->setBlockStartTrimming(newBlockStartTrimmingForSubtree);
+    auto resetBlockStartTrimming = WTF::makeScopeExit([&] {
+        layoutState->setBlockStartTrimming(blockStartTrimmingFromContainingBlock);
+    });
```

Comment:
1. there's also textbox trimming going on, maybe it's time to prefix all these (and the state API) with marginTrim
2. line clamp and textbox trimming both have their FoobarTrimmer/Updater (stack) helper classes. Maybe it's time to have one for MarginTrim? it certainly helps with scoping.

Reply:
Yeah these are both pretty nice cleanups. Will do a couple of follow up patches on top of this